### PR TITLE
Adding codebase support to firebase functions:delete

### PR DESCRIPTION
**Pull request by shettyvarun268**
### Summary
Extends `firebase functions:delete` to support deleting by **codebase** (`<codebase>`) and **codebase-qualified function** (`<codebase>:<function>`), achieving parity with `firebase deploy --only functions:...`.

### Motivation
Previously, `functions:delete` bypassed the central `parseFunctionSelector` and relied solely on raw function ID prefix matching (`f.split(/[-.]/)`). This caused:
1. Inability to delete codebases whose function exports don't share the codebase name (e.g. kit `firestore-to-bigquery` exporting `syncDocs`).
2. Inability to target a specific codebase (`functions:delete <codebase>:<function>`).
3. Accidental deletions across overlapping instance names (e.g. `kit-bq` deleting `kit-bq-staging`).

### Key Changes
1. **Reused Central Selector Parser (`functions-delete.ts`)**: Normalized `firebase.json` via `projectConfig.normalizeAndValidate()` (supporting standard codebases and Function Kit instances) and routed filters through `helper.parseFunctionSelector()`.
2. **Enhanced Matching & Boundary Safety (`functionsDeployHelper.ts`)**: Added whole-codebase matching when `!filter.idChunks` and enforced strict hyphen boundary checks to prevent substring collisions across instances.
3. **Collision Handling (`functions-delete.ts`)**: When a target matches both a codebase and a function in `default`, codebase deletion takes precedence and the CLI prints a helpful notice for the `default:<name>` workaround.

### Verification
- **Unit Tests**: Added test cases for whole-codebase matching, instance isolation, and boundary safety (`32 passing`).
- **Edge Cases**: Verified 9 edge-case scenarios in simulation.
- **Live GCP**: Verified against live Google Cloud infrastructure on project `varun-test-project-auth` for both whole-codebase and codebase-qualified function deletions.

### CODE REVIEW

File: src/commands/functions-delete.ts [hunk 1]
```diff
@@ -32,17 +33,22 @@ export const command = new Command("functions:delete [filters...]")
       return utils.reject("Must supply at least function or group name.");
     }
 
+    // Normalize project configuration to discover all registered codebases.
+    // Parse filters using parseFunctionSelector for 1:1 syntax parity with `firebase deploy --only functions:...`.
+    const config = projectConfig.normalizeAndValidate(options.config?.src?.functions);
```
[the remark this answers is not in the document]
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Added safe fallback to an empty config array ([]) when options.config?.src?.functions is not present so functions:delete continues to work outside of project directories.

File: src/deploy/functions/functionsDeployHelper.ts [hunk 2]
```diff
@@ -27,31 +27,26 @@ export function endpointMatchesAnyFilter(
...
       return false;
     }
   }
 
-  if (!filter.idChunks) {
-    // If idChunks is not provided, we match all functions.
+  // If idChunks is not provided or empty, the filter matches all functions within the targeted codebase.
+  if (!filter.idChunks || filter.idChunks.length === 0) {
     return true;
   }
```
[the remark this answers is not in the document]
Comment (shettyvarun268) [reply src/deploy/functions/functionsDeployHelper.ts:50]: Updated the logic to check (endpoint.codebase || DEFAULT_CODEBASE) === filter.codebase when !filter.idChunks, ensuring legacy functions with codebase: undefined are not incorrectly matched when targeting non-default codebases.

File: src/deploy/functions/functionsDeployHelper.spec.ts [hunk 3]
```diff
@@ -134,6 +134,42 @@ describe("functionsDeployHelper", () => {
...
+    it("should match all functions in a codebase when idChunks is not provided", () => {
+      const func1 = { ...ENDPOINT, id: "func1", codebase: "my-codebase" };
+      const func2 = { ...ENDPOINT, id: "func2", codebase: "my-codebase" };
+      const otherFunc = { ...ENDPOINT, id: "func3", codebase: "other-codebase" };
+
+      const filter: EndpointFilter = { codebase: "my-codebase" };
+      expect(helper.endpointMatchesFilter(func1, filter)).to.be.true;
+      expect(helper.endpointMatchesFilter(func2, filter)).to.be.true;
+      expect(helper.endpointMatchesFilter(otherFunc, filter)).to.be.false;
+    });
```
[the remark this answers is not in the document]
Comment (shettyvarun268) [reply src/deploy/functions/functionsDeployHelper.spec.ts:157]: Added undefinedFunc (codebase: undefined) to the unit test to ensure it is not matched by non-default codebase filters. All tests passing

File: src/commands/functions-delete.ts [hunk 4]
```diff
@@ -32,17 +33,25 @@ export const command = new Command("functions:delete [filters...]")
       return utils.reject("Must supply at least function or group name.");
     }
 
+    // Normalize project configuration to discover all registered codebases.
+    // Gracefully fall back to an empty configuration if running outside a Firebase directory or without functions.
+    // Parse filters using parseFunctionSelector for 1:1 syntax parity with `firebase deploy --only functions:...`.
+    const config = options.config?.src?.functions
+      ? projectConfig.normalizeAndValidate(options.config.src.functions)
+      : ([] as unknown as projectConfig.ValidatedConfig);
```
[the remark this answers is not in the document]
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Instead of using type assertions or dummy objects, updated parseFunctionSelector to accept ValidatedSingle[] = [] so that config cleanly falls back to [] with full static type safety.

File: src/deploy/functions/functionsDeployHelper.spec.ts [hunk 5]
```diff
@@ -135,6 +134,44 @@ describe("functionsDeployHelper", () => {
...
+
+      const filter: EndpointFilter = {
+        codebase: "kit-firestore-to-bigquery",
+      };
+
+      expect(helper.endpointMatchesFilter(instance1Func, filter)).to.be.true;
+      expect(helper.endpointMatchesFilter(instance2Func, filter)).to.be.false;
+    });
+
+    it("should not match functions with overlapping word prefixes", () => {
```
Comment (ajperel) [inline src/deploy/functions/functionsDeployHelper.spec.ts:184]: I thnik you've added support for it so maybe a case for a test like

functions:delete codebase:foo to target a specific foo when it exists in 2 codebases?

**Review by ajperel — COMMENTED** [review]
Can look more carefully later but one quick test suggestion. Nothing else jumped out at me from a quick review :)

File: src/deploy/functions/functionsDeployHelper.spec.ts (hunk 5 above)
Comment (shettyvarun268) [reply src/deploy/functions/functionsDeployHelper.spec.ts:184]: Included a unit test verifying codebase:foo targeting a specific function in codebaseA while not matching function foo in codebaseB, as well as a parser test for codebase:func selectors.

File: src/deploy/functions/functionsDeployHelper.ts [hunk 6]
```diff
@@ -27,63 +27,51 @@ export function endpointMatchesAnyFilter(
...
 /**
  * Returns true if endpoint matches the given filter.
+ * Supports filtering by codebase, exact function name, or hierarchical function group.
  */
 export function endpointMatchesFilter(endpoint: backend.Endpoint, filter: EndpointFilter): boolean {
-  // Only enforce codebase-based filtering when both the endpoint and filter provides them.
-  // This allows us to filter using idChunks across all codebases.
+  // Only enforce codebase-based filtering when both the endpoint and filter provide them.
+  // This allows us to filter using idChunks across all codebases or target a specific codebase.
   if (endpoint.codebase && filter.codebase) {
```
Comment (ajperel) [inline src/deploy/functions/functionsDeployHelper.ts]: My gemini instance thinks there is a bug here and that you need to do a

`const endpointCodebase = endpoint.codebase || DEFAULT_CODEBASE`

and compare against that. I was trying to dig into it and understand more myself and see if this is a hallucination, but I'm unable to run functions:delete on anything.

**Review by ajperel — COMMENTED** [review]
If I build and try to run this every single attempt to do `firebase functions:delete <X>` results in the error:

Error: functions.codebase must be unique but 'default' was used more than once.

I've interrigated Jeski about this and I think it's because you're on an older version of tools that doesn't yet support kits and I have a bunch of kits in my test project. But since delete update is _for_ kits we should definitely test it with kits so... maybe a first step is rebase onto main and then we go from there?

File: src/deploy/functions/functionsDeployHelper.ts (hunk 6 above)
Comment (shettyvarun268) [reply src/deploy/functions/functionsDeployHelper.ts]: Rebased the PR onto main. The kits should validate now with the experiments turned on

As for Gemini's suggestion, I think it's a hallucination here. We intentionally don't enforce DEFAULT_CODEBASE on function names so we can still match older GCP functions without codebase labels (there's an existing test for this: "should match function if backend's codebase is undefined"). We only enforce it for whole-codebase deletes.

Comment (ajperel) [reply src/deploy/functions/functionsDeployHelper.ts]: Yeah... I was a bit skeptical myself since it was encouraging changing this and a unit test together that you didn't touch. But... I _think_ it might be a real thing. 

The claim is that if you have a function without a codebase like `helloWorld` that it would be deleted if you specify _any_ codebase like `default:helloWorld` (which I'd expect to work) but also `my-new-codebase:helloWorld`.  That may have been the old behavior but it doesn't really make sense right? We'll put _any_ function with no codebase into the default codebase right?

So I think that if it has no codebase then `helloWorld` or `default:helloWorld` should match but `my-function-kit:helloWorld` probably shouldn't match this legacy function?

Can you look into it a bit more?

**Review by ajperel — COMMENTED** [review]
I think we do have a behavior change here. I have in my test setup a function node-functions:v2-helloWorldNode. Before this change

`firebase functions:delete v2-helloWorldNode` would prompt me to delete `node-functions:v2-helloWorldNode`. Now it finds nothing. However before the opposite was true `node-functions:v2-helloWorldNode` didn't work and now it does.

I think the new functionality is strictly more powerful, and more inline with how only/except work. So it's probably a good change. But it _is_ a larger behavior change. So I think we should rediscuss with folks on if we have to be more careful with how we roll this out.

File: src/deploy/functions/functionsDeployHelper.ts (hunk 6 above)
Comment (shettyvarun268) [reply src/deploy/functions/functionsDeployHelper.ts]: Thanks for pointing this out.

1. Changed the behavior regarding legacy functions which did not have codebases attached to them. Now if the entrypoint.codebase == undefined, it will behave as if it has the default code base attached to it.

2. After our discussion, decided to keep the old behavior in the delete functionality where running bare functions:delete foo (without a codebase prefix) continues to match across all codebases as it previously did. This would avoid us doing a breaking change and can be changed in the (future when we decide to actually push through a breaking change and package this in.

File: src/commands/functions-delete.ts [hunk 7]
```diff
@@ -67,6 +78,30 @@ export const command = new Command("functions:delete [filters...]")
...
 
+    // Inform the user when a name collision exists between a codebase name and a function name in default.
+    // Codebase deletion takes precedence by design, but we provide the explicit 'default:<name>' workaround.
+    const codebaseNames = helper.getCodebasesFromConfig(config);
+    const defaultEndpoints = backend
+      .allEndpoints(existingBackend)
+      .filter((ep) => !ep.codebase || ep.codebase === projectConfig.DEFAULT_CODEBASE);
+
+    for (const f of filters) {
+      if (!f.includes(":") && codebaseNames.includes(f)) {
```
Comment (inlined) [inline src/commands/functions-delete.ts]: Always prefer early return/continue/break instead of deep nesting.

File: src/commands/functions-delete.ts [hunk 8]
```diff
@@ -67,6 +78,30 @@ export const command = new Command("functions:delete [filters...]")
       );
     }
 
+    // Inform the user when a name collision exists between a codebase name and a function name in default.
+    // Codebase deletion takes precedence by design, but we provide the explicit 'default:<name>' workaround.
+    const codebaseNames = helper.getCodebasesFromConfig(config);
+    const defaultEndpoints = backend
```
Comment (inlined) [inline src/commands/functions-delete.ts]: Technically there are tools made so you don't have to know what goes in teh backend struct. E.g. here you can say `backend.matchingBackend(haveBackend, (ep) => !ep.codebase || ..)`

File: src/commands/functions-delete.ts [hunk 9]
```diff
@@ -67,6 +78,30 @@ export const command = new Command("functions:delete [filters...]")
...
+      .allEndpoints(existingBackend)
+      .filter((ep) => !ep.codebase || ep.codebase === projectConfig.DEFAULT_CODEBASE);
+
+    for (const f of filters) {
+      if (!f.includes(":") && codebaseNames.includes(f)) {
+        const hasDefaultCollision = defaultEndpoints.some(
```
Comment (inlined) [inline src/commands/functions-delete.ts]: Are we only handling collisions in the default codebase? Why?

File: src/commands/functions-delete.ts [hunk 10]
```diff
@@ -67,6 +78,30 @@ export const command = new Command("functions:delete [filters...]")
       );
     }
 
+    // Inform the user when a name collision exists between a codebase name and a function name in default.
+    // Codebase deletion takes precedence by design, but we provide the explicit 'default:<name>' workaround.
+    const codebaseNames = helper.getCodebasesFromConfig(config);
```
Comment (inlined) [inline src/commands/functions-delete.ts]: I don't understand why we need the codebases from the directory. We have the ones in prod. Codebases work with  multiple repos in a project and this would mean that you can only delete functions from the repository you're in.

File: src/commands/functions-delete.ts [hunk 11]
```diff
@@ -32,17 +33,27 @@ export const command = new Command("functions:delete [filters...]")
...
+    const [firebaseConfig, existingBackend] = await Promise.all([
       functionsConfig.getFirebaseConfig(options),
       backend.existingBackend(context),
     ]);
     await backend.checkAvailability(context, /* want=*/ backend.empty());
-    const appEngineLocation = functionsConfig.getAppEngineLocation(config);
+    const appEngineLocation = functionsConfig.getAppEngineLocation(firebaseConfig);
 
     if (options.region) {
       existingBackend.endpoints = { [options.region]: existingBackend.endpoints[options.region] };
```
Comment (inlined) [inline src/commands/functions-delete.ts]: Should probably be `backend.filter(existingBackend, (ep) => ep.region === options.region);`

File: src/commands/functions-delete.ts [hunk 12]
```diff
@@ -32,17 +33,27 @@ export const command = new Command("functions:delete [filters...]")
...
     }
 
+    // Normalize project configuration to discover all registered codebases.
+    // Gracefully fall back to an empty configuration if running outside a Firebase directory or without functions.
+    // Parse filters using parseFunctionSelector for 1:1 syntax parity with `firebase deploy --only functions:...`.
+    const config = options.config?.src?.functions
+      ? projectConfig.normalizeAndValidate(options.config.src.functions)
+      : [];
+    const parsedFilters = filters.flatMap((f) =>
+      helper.parseFunctionSelector(f, config, /* defaultCodebase= */ undefined),
```
Comment (inlined) [inline src/commands/functions-delete.ts]: It seems like half of your code is just making this helper usable in this one case, but it's changing code used elsewhere to have a nonsense case (choosing a custom name for the default codebase) no? I wonder if it's easier to just make parseFunctionSelector just always emit two filters when it's abmiguous? The API is already designed to support returning an array of filters but never does... I wonder why that was removed.

I think you could do something like that to get an array of possible EndpointFilters for your selectors. Then you iterate over that array of arrays. For each sub array if there is zero matches you warn that the selector had no match. If there is one match you accept it. If there is more than one match you warn about the ambiguity.

**Review by inlined — CHANGES_REQUESTED** [review]

File: src/commands/functions-delete.ts (hunk 7 above)
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Done. Inverted the condition to use continue

File: src/commands/functions-delete.ts (hunk 8 above)
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Good point. Refactored to use backend.allEndpoints and backend.matchingBackend helpers instead of poking at the struct directly.

File: src/commands/functions-delete.ts (hunk 9 above)
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Updated to check for collisions against all active endpoints and output the exact <codebase>:<func> prefix in the notice.

File: src/commands/functions-delete.ts (hunk 10 above)
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: We now discover active codebases directly from live prod endpoints (existingBackend) as well as local config.

File: src/commands/functions-delete.ts (hunk 12 above)
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Agreed, that parameter added unnecessary clutter. Removed the custom defaultCodebase arg from parseFunctionSelector completely so its signature remains identical to main. For bare selectors without a colon, we simply omit the codebase restriction in functions-delete.ts so they match across all codebases.

File: src/commands/functions-delete.ts (hunk 11 above)
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Updated to use backend.matchingBackend(existingBackend, (ep) => ep.region === options.region).

File: src/commands/functions-delete.ts (hunk 10 above)
Comment (ajperel) [reply src/commands/functions-delete.ts]: shouldn't it _only_ look at live endpoints? If it's in the config but not live there is nothing to delete.

Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Noted. We now discover the codebases by looking only through the live endpoints.

File: src/commands/functions-delete.ts [hunk 13]
```diff
@@ -34,19 +35,45 @@ export const command = new Command("functions:delete [filters...]")
...
+    const activeCodebases = [
+      ...new Set(
+        backend
+          .allEndpoints(existingBackend)
+          .map((ep) => ep.codebase || projectConfig.DEFAULT_CODEBASE),
+      ),
+    ];
+    const liveCodebasesConfig = activeCodebases.map((codebase) => ({ source: "", codebase }));
+
+    const parsedFilters = filters.flatMap((f) => {
```
Comment (inlined) [inline src/commands/functions-delete.ts]: This section can really use a comment. Basically, if I understand correctly, you're using the list of live codebases to make every filter that targets the default codebase not include a codebase selector?

File: src/commands/functions-delete.ts [hunk 14]
```diff
@@ -67,6 +94,31 @@ export const command = new Command("functions:delete [filters...]")
...
+    // Codebase deletion takes precedence by design, but we provide the explicit '<codebase>:<name>' workaround.
+    const allEndpoints = backend.allEndpoints(existingBackend);
+    for (const f of filters) {
+      if (f.includes(":") || !activeCodebases.includes(f)) {
+        continue;
+      }
+      const matchingEndpoints = allEndpoints.filter(
+        (ep) => ep.id === f || ep.id.startsWith(`${f}-`),
+      );
+      if (matchingEndpoints.length > 0) {
```
Comment (inlined) [inline src/commands/functions-delete.ts]: I think this section could really use comments. I had to write this down.

101: // If the filter is codebase scoped or doesn't match a codebase there is no collision
104: // This does match a codebase, so if it matches by IDs there is a collision

**Review by inlined — APPROVED** [review]
It would be great if there was any testing of this sort of logic since it reads pretty densely. E.g. maybe pull some of the helpers like "detectCodebaseAndIdCollisions" into a helper and test it.

File: src/commands/functions-delete.ts (hunk 14 above)
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Done. Left a general comment in the PR.

File: src/commands/functions-delete.ts (hunk 13 above)
Comment (shettyvarun268) [reply src/commands/functions-delete.ts]: Done. Left a general comment in the PR.

**Comment by shettyvarun268** [comment]
Thank you so much for the feedback @inlined. Made the following changes to the code:

1. Extracted parseDeleteFilters and detectCodebaseAndIdCollisions into src/deploy/functions/functionsDeployHelper.ts.
2. Added detailed JSDoc comments explaining the unqualified selector logic and the codebase vs ID collision logic.
3. Added unit test suites covering both helper functions in src/deploy/functions/functionsDeployHelper.spec.ts.
